data/blog/improving-nodejs-loader-performance.md

```diff
@@ -0,0 +1,180 @@
+  if (mainPath && mainPath.endsWith('.mjs')) { return true; }
+  if (!mainPath || mainPath.endsWith('.cjs')) { return false; }
+
+  const pkg = readPackageScope(mainPath);
+  switch (pkg.data?.type) {
+    case 'module':
+      return true;
+    case 'commonjs':
+      return false;
+    default: { // No package.json or no `type` field.
+      return false;
+    }
+  }
+}
+```
+
+`readPackageScope` traverses the directory tree upwards until it finds a `package.json` file. Prior to the optimizations done on this post, `readPackageScope` calls an internal version of `fs.readFileSync` until it finds a `package.json` file. This synchronous call makes a filesystem operation and communicates with Node.js C++ layer. This operation has performance bottlenecks depending on the value/type it returns because of the cost of serialization/deserialization of data. This is why we want to avoid calling `readPackage` a.k.a. `fs.readFileSync` inside `readPackageScope` as much as possible.
+
+## How does Node.js parses `package.json`?
+
+By default, `readPackage` calls an internal version `fs.readFileSync` to read the `package.json` file. This synchronous call returns a string from Node.js C++ layer, which later gets parsed using V8's `JSON.parse()` method. Depending on the validity of this JSON, Node.js checks and creates an object that's required for the remaining of the loaders to perform. These fields are `pkg.name`, `pkg.main`, `pkg.exports`, `pkg.imports` and `pkg.type`. If the JSON has faulty syntax, Node.js will throw an error and exit the process.
+
+The output of this function is later cached at an internal `Map` to avoid calling `readPackageScope` again for the same path. This cache is stored for the rest of the process lifetime.
+
+## Usage of `package.json` fields and the reader
+
+Before we dive into what optimizations we can do, let's see how Node.js uses these fields. The common use cases in Node.js codebase for parsing and re-using `package.json` fields are:
+
+- `pkg.exports` and `pkg.imports` are used to resolve different modules according to your input.
+- `pkg.main` is used to resolve the entry point of the application.
+- `pkg.type` is used to resolve the module format of the file.
+- `pkg.name` is used if there is a self referencing require/import.
+
+Additionally, Node.js supports an experimental version of `Subresource Integrity` checkwhich uses the result of this package.json to validate the integrity of the file.
+
+The most important usage is that, for every `require/import` call, Node.js needs to know the module format of the file. For example, if the user require's a NPM module that uses ESM on a CommonJS (CJS) application, Node.js will need to parse the `package.json` file of that module and throw an error if the NPM package is ESM.
+
+Because of all of these calls and usages across ESM and CJS loaders, `package.json` reader is one of the most important parts of Node.js loader implementation.
```

Comment:
```suggestion
Because of all of these calls and usages across ESM and CJS loaders, `package.json` reader is one of the most important parts of the Node.js loader implementation.
```